pkg/pipelines/imagerepo/imagerepo.go

```diff
@@ -0,0 +1,91 @@
+package imagerepo
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/openshift/odo/pkg/pipelines/meta"
+	"github.com/openshift/odo/pkg/pipelines/namespaces"
+	"github.com/openshift/odo/pkg/pipelines/roles"
+
+	"github.com/openshift/odo/pkg/pipelines/config"
+
+	res "github.com/openshift/odo/pkg/pipelines/resources"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+// ValidateImageRepo validates the input image repo.  It determines if it is
+// for internal registry and prepend internal registry hostname if neccessary.
+func ValidateImageRepo(imageRepo, registryURL string) (bool, string, error) {
+	components := strings.Split(imageRepo, "/")
+
+	// repo url has minimum of 2 components
+	if len(components) < 2 {
+		return false, "", imageRepoValidationErrors(imageRepo)
+	}
+
+	for _, v := range components {
+		// check for empty components
+		if strings.TrimSpace(v) == "" {
+			return false, "", imageRepoValidationErrors(imageRepo)
+		}
+		// check for white spaces
+		if len(v) > len(strings.TrimSpace(v)) {
+			return false, "", imageRepoValidationErrors(imageRepo)
+		}
+	}
+
+	if len(components) == 2 {
+		if components[0] == "docker.io" || components[0] == "quay.io" {
+			// we recognize docker.io and quay.io.  It is missing one component
+			return false, "", imageRepoValidationErrors(imageRepo)
+		}
+		// We have format like <project>/<app> which is an internal registry.
+		// We prepend the internal registry hostname.
+		return true, registryURL + "/" + imageRepo, nil
+	}
+
+	// Check the first component to see if it is an internal registry
+	if len(components) == 3 {
+		return components[0] == registryURL, imageRepo, nil
+	}
+
+	// > 3 components.  invalid repo
+	return false, "", imageRepoValidationErrors(imageRepo)
+}
+
+func imageRepoValidationErrors(imageRepo string) error {
+	return fmt.Errorf("failed to parse image repo:%s, expected image repository in the form <registry>/<username>/<repository> or <project>/<app> for internal registry", imageRepo)
+}
+
+func CreateInternalRegistryResources(cicdEnv *config.Environment, sa *corev1.ServiceAccount, imageRepo string) (res.Resources, error) {
+
+	// Provide access to service account for using internal registry
+	namespace := strings.Split(imageRepo, "/")[1]
+
+	clientSet, err := namespaces.GetClientSet()
+	if err != nil {
+		return nil, err
+	}
+	namespaceExists, err := namespaces.Exists(clientSet, namespace)
+	if err != nil {
+		return nil, err
+	}
+	resources := res.Resources{}
+
+	if !namespaceExists {
```

Comment:
Should we check if the namespace/environment exists in the manifest file? If it's an existing env we don't have to create the namespace and rolebinding files

Reply:
I thought  it was a good nice idea at first.   I actually went down the  path to implement it but I decided to back it out.   Manifest only records namespaces of environments.  The chance of image repo projects/namespaces overlap with environment names is slim.    In order to detect duplicated namespaces in generated resources, we will have to reply on the yamls on the filesystem to be present (especially during the "service add" operation)  and will have to parse them.   Generating overlapping namespace yamls does not harm anything and it may not be a bad alternative.   Also, the current way of generating service resources really make deletion problematic.   We may want  to re-think how "service add" should generate resources.   (  E.g.  service a and b both has internal repo name space "proj".   If I delete service a, I can't delete proj namespace because service b still have reference to it.

Reply:
Yeah. I didn't consider the service deletion scenario. It makes sense not to have services dependent on each other